discretize/base/base_mesh.py

```diff
@@ -2246,6 +2657,285 @@ def get_edge_inner_product_deriv(
+            returns a (``n_edges``, ``n_params``) :class:`scipy.sparse.csr_matrix`.
+        """
+        raise NotImplementedError(
+            f"get_edge_inner_product_surface_deriv not implemented for {type(self)}"
+        )
+
+    def get_face_inner_product_surface_deriv(
+        self,
+        model,
+        invert_model=False,
+        invert_matrix=False,
+        **kwargs,
+    ):
+        r"""Get a function handle to multiply a vector with derivative of face inner product surface matrix (or its inverse).
+
+        Let :math:`\mathbf{M}(\mathbf{m})` be the face inner product surface matrix
+        constructed with a set of diagnostic property parameters :math:`\mathbf{m}`
+        (or its inverse) defined on mesh faces. **get_face_inner_product_surface_deriv**
+        constructs a function handle
+
+        .. math::
+            \mathbf{F}(\mathbf{u}) = \mathbf{u}^T \, \frac{\partial \mathbf{M}(\mathbf{m})}{\partial \mathbf{m}}
+
+        which accepts any numpy.array :math:`\mathbf{u}` of shape (n_faces,). That is,
+        **get_face_inner_product_surface_deriv** constructs a function handle for computing
+        the dot product between a vector :math:`\mathbf{u}` and the derivative of the
+        face inner product surface matrix (or its inverse) with respect to the property parameters.
+        When computed, :math:`\mathbf{F}(\mathbf{u})` returns a ``scipy.sparse.csr_matrix``
+        of shape (n_faces, n_param).
+
+        The function handle can only be created for isotropic diagnostic properties.
+
+        Parameters
+        ----------
+        model : (n_faces, ) numpy.ndarray
+            Parameters defining the diagnostic property values for every face in the mesh.
+        invert_model : bool, optional
+            The inverse of *model* is used as the diagnostic property.
+        invert_matrix : bool, optional
+            Returns the inverse of the inner product surface matrix.
+
+        Returns
+        -------
+        function
+            The function handle :math:`\mathbf{F}(\mathbf{u})` which accepts a
+            (``n_faces``) :class:`numpy.ndarray` :math:`\mathbf{u}`. The function
+            returns a (``n_faces``, ``n_params``) :class:`scipy.sparse.csr_matrix`.
+        """
+        try:
+            if model is None:
+                tensorType = -1
+            elif is_scalar(model):
+                tensorType = 0
+            elif model.size == self.nF:
+                tensorType = 1
+            else:
+                raise Exception(
+                    "Unexpected shape of tensor: {}".format(model.shape),
+                    "Must be scalar or have length equal to total number of faces.",
+                )
+
+            A = sdiag(self.face_areas)
+
+        except NotImplementedError:
+            raise NotImplementedError(
+                f"get_face_inner_product_surface_deriv not implemented for {type(self)}"
+            )
```

Comment:
again put the test for if `face_areas` is implemented seperate fom the rest of the code so it is clear that that is what you are catching the exception from.